File: data/Generator/templates/react-app/src/entities/_entity/_Entity.js
{{={= =}=}}

import uuidv4 from 'uuid/v4'

export default class {= entity.name =} {
  _data = {}

  constructor (data = {}) {
    this._data = {
      id: data.id || uuidv4(),
      {=# entity.fields =}
      {= name =}: data.{= name =},
      {=/ entity.fields =}
    }
  }

  get id () {
    return this._data.id
  }

  {=# entity.fields =}
  get {= name =} () {
    return this._data.{= name =}
  }
  {=/ entity.fields =}

  toData () {
    return this._data
  }
}

File: waspc/src/Wasp/Wasp/Entity.hs
module Wasp.Wasp.Entity
  ( Entity (..),
    Field (..),
    FieldType (..),
    Scalar (..),
    Composite (..),
  )
where

import Data.Aeson (ToJSON (..), object, (.=))
import qualified Wasp.Psl.Ast.Model

data Entity = Entity
  { _name :: !String,
    _fields :: ![Field],
    _pslModelBody :: !Wasp.Psl.Ast.Model.Body
  }
  deriving (Show, Eq)

data Field = Field
  { _fieldName :: !String,
    _fieldType :: !FieldType
  }
  deriving (Show, Eq)

data FieldType = FieldTypeScalar Scalar | FieldTypeComposite Composite
  deriving (Show, Eq)

data Composite = Optional Scalar | List Scalar
  deriving (Show, Eq)

data Scalar
  = String
  | Boolean
  | Int
  | BigInt
  | Float
  | Decimal
  | DateTime
  | Json
  | Bytes
  | -- | Name of the user-defined type.
    -- This could be another entity, or maybe an enum,
    -- we don't know here yet.
    UserType String
  | Unsupported String
  deriving (Show, Eq)

instance ToJSON Entity where
  toJSON entity =
    object
      [ "name" .= _name entity,
        "fields" .= show (_fields entity),
        "pslModelBody" .= show (_pslModelBody entity)
      ]
